File: benchmarks/pandas/bench_dataframe_shift_diff.py
"""
Benchmark: DataFrame.shift / DataFrame.diff — shift and diff on a 50k-row DataFrame.
Outputs JSON: {"function": "dataframe_shift_diff", "mean_ms": ..., "iterations": ..., "total_ms": ...}
"""
import json
import time
import numpy as np
import pandas as pd

SIZE = 50_000
WARMUP = 5
ITERATIONS = 50

rng = np.random.default_rng(42)
df = pd.DataFrame({
    "a": np.arange(SIZE) * 1.5,
    "b": np.sin(np.arange(SIZE) * 0.01) * 100,
    "c": np.arange(SIZE) % 200,
})

for _ in range(WARMUP):
    df.shift(1)
    df.diff(1)

start = time.perf_counter()
for _ in range(ITERATIONS):
    df.shift(1)
    df.diff(1)
total = (time.perf_counter() - start) * 1000

print(json.dumps({
    "function": "dataframe_shift_diff",
    "mean_ms": total / ITERATIONS,
    "iterations": ITERATIONS,
    "total_ms": total,
}))

File: benchmarks/tsb/bench_dataframe_shift_diff.ts
/**
 * Benchmark: dataFrameShift / dataFrameDiff — shift and diff on a 50k-row DataFrame.
 * Outputs JSON: {"function": "dataframe_shift_diff", "mean_ms": ..., "iterations": ..., "total_ms": ...}
 */
import { DataFrame, dataFrameShift, dataFrameDiff } from "../../src/index.ts";

const SIZE = 50_000;
const WARMUP = 5;
const ITERATIONS = 50;

const df = new DataFrame({
  a: Array.from({ length: SIZE }, (_, i) => i * 1.5),
  b: Array.from({ length: SIZE }, (_, i) => Math.sin(i * 0.01) * 100),
  c: Array.from({ length: SIZE }, (_, i) => i % 200),
});

for (let i = 0; i < WARMUP; i++) {
  dataFrameShift(df, 1);
  dataFrameDiff(df, 1);
}

const start = performance.now();
for (let i = 0; i < ITERATIONS; i++) {
  dataFrameShift(df, 1);
  dataFrameDiff(df, 1);
}
const total = performance.now() - start;

console.log(
  JSON.stringify({
    function: "dataframe_shift_diff",
    mean_ms: total / ITERATIONS,
    iterations: ITERATIONS,
    total_ms: total,
  }),
);
